bcgov/cas-registration
# 🏗️ feat: role protected app routes

Addresses [Ticket 265](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/265)

## 🚧 Changes

    - Modify `client/app/api/auth/[...nextauth]/route.ts`
    - Modify `client/middlewares/withAuthorization.tsx`
    - Add protected routes `client\app\(authenticated)/`:
        - `becid/industry_user`
        - `becid/industry_user_admin`
        - `idir/cas_admin`
        - `idir/cas_analyst`

  

## 🚀 Impact:
   - App route protection implementation uses app directory to structure protected page routes based on NextAuth identity provider and user role as folders.
   - Each protected page has its own folder structure based on NextAuth JWT identity provider, NextAuth JWT user role, and a dashboard folder (used for dynamic breadcrumbs);   e.g. `app/(authenticated)/idir/cas_ admin/dashboard/operations/Page.tsx`
   - Access control logic is managed using Next.js middleware and NextAuth.js authentication JWT token.  The middleware intercepts requests, checks for a valid user session, and extracts user information from the JWT token. Based on JWT properties of identity_provider and role, the middleware  dynamically rewrites the request URL base request to the appropriate folder structure e.g. `http://localhost:3000/dashboard/operations` request maps to `http://localhost:3000/idir/cas_admin/dashboard/operations` for user `idir\cas_admin`
   - Using NextAuth JWT properties based static app directory to structure protected page routes and middleware logic to build dynamic responses allows for separation of concerns, maintainability and flexibility to delete or add roles without modifying existing access control logic.
 

## 📝 Notes:
- Using dynamic route creation based on JWT token properties attempts to align access control with the user's identity and role attributes which can then accommodate changes to user roles without requiring modifying existing access control logic. Static app directory protected routes can be added or removed without making any other changes. One caveat is that page routes may repeat e.g. `app/(authenticated)/idir/cas_admin/dashboard/operations` and  e.g. `app/(authenticated)/idir/cas_analyst/dashboard/operations`; so, common components are used to shared between protected page routes in an attempt to reduce repetition.

## 🛠️ To Do:
- IDIR role is hard coded to `cas_admin`, need to implement the DB lookup within `client/app/api/auth/[...nextauth]/route.ts`
- For outstanding FE PRs the legacy file structure remain but, new features will need to be integrated into protected routes as required.
- The static app protected routes are not finalized
 

##  🔬 Testing:

1. Open a terminal in cas-registration\bc_obps directory
2. Follow the [developer environment setup](https://github.com/bcgov/cas-registration/blob/develop/docs/developer-environment-setup.md#backend-environment-setup) to start the API
3.  From terminal command, start the app development server:
   ```
cd client && yarn dev
   ```
4. In your web browser, navigate to `http://localhost:3000`
5. You should be directed to the anonymous landing page
![image](https://github.com/bcgov/cas-registration/assets/120038448/bd8166b1-4f19-4993-82dc-0b3340d625e7)
6. Click "Program Administrator Log In"
7. You should be directed to IDIR login screen
![image](https://github.com/bcgov/cas-registration/assets/120038448/2e29ad50-559e-4c51-a756-c7953b9e17a8)
8. Login using your IDIR account
9. You should be redirected to the idir/cas_admin dashboard
![image](https://github.com/bcgov/cas-registration/assets/120038448/d6c62c5b-86e4-4f7b-8a08-1a03ee7bd27b)
10. Click your IDIR login name in the right hand header
11. Your user session information should display
12. Your Role should reflect the "hard-code" `cas_admin` and you IP should be `idir`
13. Click the breadcrumb "Dashboard"
14. You should be redirected to the `idir/cas_admin` dashboard
15. Click all dashboard navigation tiles
16. You should be directed to correct route page
17. Modify the browser URL to an app path accessible to `idir\cas_admin` e.g. `http://localhost:3000/idir/cas_admin/dashboard/operations`
18. You should be redirected to URL:  `http://localhost:3000/dashboard/operations` retaining the expected breadcrumb structure
19. Modify the browser URL to an app path **NOT** accessible to `idir\cas_admin` e.g.  `http://localhost:3000/idir/cas_analyst/dashboard/operations`
20. You should be redirected to "Page not found"
